lib/smart_proxy_remote_execution_ssh/api.rb

```diff
@@ -4,6 +4,120 @@ class Api < ::Sinatra::Base
       get "/pubkey" do
         File.read(Ssh.public_key_file)
       end
+      post "/session" do
+        if env["HTTP_CONNECTION"] != "upgrade" or env["HTTP_UPGRADE"] != "raw"
+          return [ 401, "Invalid request: /ssh/session requires connection upgrade to 'raw'" ]
+        end
+        socket = nil
+        if env['rack.hijack?']
+          env['rack.hijack'].call
+          socket = env['rack.hijack_io']
+        end
+        if !socket
+          return [ 501, "Internal error: request hijacking not available" ]
+        end
+
+        params = MultiJson.load(env["rack.input"].read)
+        key_file = Proxy::RemoteExecution::Ssh.private_key_file
+
+        methods = %w(publickey)
+        methods.unshift('password') if params["ssh_password"]
+
+        ssh_options = { }
+        ssh_options[:port] = params["ssh_port"] if params["ssh_port"]
+        ssh_options[:keys] = [ key_file ] if key_file
+        ssh_options[:password] = params["ssh_password"] if params["ssh_password"]
+        ssh_options[:passphrase] = params[:ssh_key_passphrase] if params[:ssh_key_passphrase]
+        ssh_options[:keys_only] = true
+        ssh_options[:auth_methods] = methods
+        ssh_options[:verify_host_key] = :accept_new_or_local_tunnel
+        ssh_options[:number_of_password_prompts] = 1
+
+        ssh_on_socket(socket, params["command"], params["ssh_user"], params["hostname"], ssh_options)
+        101
+      end
+
+      def ssh_on_socket(socket, command, ssh_user, host, ssh_options)
+        started = false
+        err_buf = ""
+        socket.extend(Net::SSH::BufferedIo)
+
+        send_start = -> {
+          if !started
+            started = true
+            socket.enqueue("Status: 101\r\n")
+            socket.enqueue("Connection: upgrade\r\n")
+            socket.enqueue("Upgrade: raw\r\n")
+            socket.enqueue("\r\n")
+          end
+        }
+
+        send_error = -> (code, msg) {
+          socket.enqueue("Status: #{code}\r\n")
+          socket.enqueue("Connection: close\r\n")
+          socket.enqueue("\r\n")
+          socket.enqueue(msg)
+        }
+
+        begin
+          Net::SSH.start(host, ssh_user, ssh_options) do |ssh|
+            channel = ssh.open_channel do |ch|
+              ch.exec(command) do |ch, success|
+                raise "could not execute command" unless success
+
+                ssh.listen_to(socket)
+
+                ch.on_process do
+                  if socket.available > 0
+                    ch.send_data(socket.read_available)
```

Comment:
This could be rewritten as
```ruby
ch.send_data(socket.read_available) if socket.available.positive?
```

Reply:
For this particular case, I've tried the modifier way and it looks a bit weird. The standard way reads a bit better here